src/ripple/protocol/impl/AccountID.cpp

```diff
@@ -17,17 +17,95 @@
 */
 //==============================================================================
 
+#include <ripple/basics/hardened_hash.h>
+#include <ripple/basics/spinlock.h>
 #include <ripple/protocol/AccountID.h>
 #include <ripple/protocol/PublicKey.h>
 #include <ripple/protocol/digest.h>
 #include <ripple/protocol/tokens.h>
+#include <array>
 #include <cstring>
+#include <mutex>
 
 namespace ripple {
 
+namespace detail {
+
+/** Caches the base58 representations of AccountIDs */
+class AccountIdCache
+{
+private:
+    struct CachedAccountID
+    {
+        AccountID id;
+        char encoding[40] = {0};
+    };
+
+    // The actual cache
+    std::vector<CachedAccountID> cache_;
+
+    // We use a hash function designed to resist algorithmic complexity attacks
+    hardened_hash<> hasher_;
+
+    // 64 spinlocks, packed into a single 64-bit value
+    std::atomic<std::uint64_t> locks_ = 0;
+
+public:
+    AccountIdCache(std::size_t count) : cache_(count)
+    {
+        // This is non-binding, but we try to avoid wasting memory that
+        // is caused by overallocation.
+        cache_.shrink_to_fit();
+    }
+
+    std::string
+    toBase58(AccountID const& id)
+    {
+        auto const index = hasher_(id) % cache_.size();
+
+        packed_spinlock sl(locks_, index % 64);
+
+        {
+            std::lock_guard lock(sl);
+
+            // The check against the first character of the encoding ensures
+            // that we don't mishandle the case of the all-zero account:
+            if (cache_[index].encoding[0] != 0 && cache_[index].id == id)
+                return cache_[index].encoding;
+        }
+
+        auto ret =
+            encodeBase58Token(TokenType::AccountID, id.data(), id.size());
+
+        assert(ret.size() <= 38);
+
+        {
+            std::lock_guard lock(sl);
+            cache_[index].id = id;
+            std::strcpy(cache_[index].encoding, ret.c_str());
```

Comment:
I know there's an `assert` a couple of lines up, but `strcpy` always raises my hackles. Would it hurt to use `strncpy`?

Reply:
It wouldn't hurt. I'll make a follow-up fix for this, if you don't mind merging this as is.